### PR TITLE
Modification of readTree and writeTree methods to preserve node labels during conversion between 'phylo' and 'TreeMan' classes.

### PR DESCRIPTION
**Pull request by T-Heide**
Please see descriptions in the commit messages for details.

### CODE REVIEW

File: R/cnvrt-methods.R [hunk 1]
```diff
@@ -50,7 +50,16 @@ setAs(from="phylo", to="TreeMan", def=function(from, to) {
 #' tree <- as(tree, 'TreeMan')
 setAs(from="TreeMan", to="phylo", def=function(from, to) {
   temp_file <- tempfile(pattern="temp_tree", fileext=".tre")
-  writeTree(from, file=temp_file)
+  
+  # use a node label function function that writes the node labels during the 
+  # conversion, if they are non default.
+  if (all(from["nds"] == paste0("n", seq_along(from["nds"])))) {
+    node_labeller <- function(nd) { return(NULL) } # no node labels
```
Comment (DomBennett) [inline R/cnvrt-methods.R:58]: When I started the project, I chose camel-case for for function names. Not standard for most R projects now.

```suggestion
    nodeLabeller <- function(nd) { return(NULL) } # no node labels 
```

File: R/cnvrt-methods.R [hunk 2]
```diff
@@ -50,7 +50,16 @@ setAs(from="phylo", to="TreeMan", def=function(from, to) {
...
 setAs(from="TreeMan", to="phylo", def=function(from, to) {
   temp_file <- tempfile(pattern="temp_tree", fileext=".tre")
-  writeTree(from, file=temp_file)
+  
+  # use a node label function function that writes the node labels during the 
+  # conversion, if they are non default.
+  if (all(from["nds"] == paste0("n", seq_along(from["nds"])))) {
+    node_labeller <- function(nd) { return(NULL) } # no node labels 
+  } else {
+    node_labeller <- function(n) { print(n); n[['id']] } # writes node labels
```
Comment (DomBennett) [inline R/cnvrt-methods.R]:
```suggestion
    nodeLabeller <- function(n) { n[['id']] } # writes node labels 
```
I don't thinking printing is necessary.

File: R/cnvrt-methods.R [hunk 3]
```diff
@@ -50,7 +50,16 @@ setAs(from="phylo", to="TreeMan", def=function(from, to) {
...
+  
+  # use a node label function function that writes the node labels during the 
+  # conversion, if they are non default.
+  if (all(from["nds"] == paste0("n", seq_along(from["nds"])))) {
+    node_labeller <- function(nd) { return(NULL) } # no node labels 
+  } else {
+    node_labeller <- function(n) { print(n); n[['id']] } # writes node labels 
+  }
+
+  writeTree(from, file=temp_file, ndLabels = node_labeller)
```
Comment (DomBennett) [inline R/cnvrt-methods.R]:
```suggestion
  writeTree(from, file=temp_file, ndLabels = nodeLabeller)
```

File: R/read-write-methods.R [hunk 4]
```diff
@@ -199,7 +199,10 @@ readTree <- function(file=NULL, text=NULL, spcl_slt_nm='Unknown', wndmtrx=FALSE,
   other <- rep(NA, length(ids))
   intnds <- 1:length(ids) %in% prinds
   other[intnds] <- ids[intnds]
-  ids[intnds] <- paste0('n', which(intnds))
+  # overwrite internal node ids only if at least one is malformed
+  if (any(grepl('[^a-zA-Z_0-9]', ids[intnds]))){
+    ids[intnds] <- paste0('n', which(intnds))
+  }
```
Comment (DomBennett) [inline R/read-write-methods.R:208]:
```suggestion
  } else {
      # if using internal node ids as ids, no need for special slot
      other[intnds] <- NA
  }
```
No need to have a special node slot if the same information is contained in the IDs themsevles. TreeMan objects are memory-heavy enough as it is.

File: tests/testthat/test-rw-methods.R [hunk 5]
```diff
@@ -25,13 +25,21 @@ test_that('readTree([w/o spans]) works', {
...
+test_that('readTree([w/ malformed node labels]) works', {
   text <- "(A,B,(C,D)0.9)0.8;"
   tree <- readTree(text=text, spcl_slt_nm='bootstrap',
                    wndmtrx=sample(c(TRUE, FALSE), 1))
   expect_that(tree['ntips'], equals(4))
   expect_false(tree['wspn'])
 })
+test_that('readTree([w/ node labels]) works', {
+  text <- "(A,B,(C,D)N1)N2;"
+  tree <- readTree(text=text, spcl_slt_nm='bootstrap',
```
Comment (DomBennett) [inline tests/testthat/test-rw-methods.R:46]: Special slot name "bootstrap" should not be needed.

File: tests/testthat/test-rw-methods.R [hunk 6]
```diff
@@ -25,13 +25,21 @@ test_that('readTree([w/o spans]) works', {
   expect_that(tree['ntips'], equals(4))
   expect_false(tree['wspn'])
 })
-test_that('readTree([w/ node labels]) works', {
+test_that('readTree([w/ malformed node labels]) works', {
```
Comment (DomBennett) [inline tests/testthat/test-rw-methods.R]: Can "malformed" be "non-id"?

File: tests/testthat/test-rw-methods.R [hunk 7]
```diff
@@ -25,13 +25,21 @@ test_that('readTree([w/o spans]) works', {
...
   text <- "(A,B,(C,D)0.9)0.8;"
   tree <- readTree(text=text, spcl_slt_nm='bootstrap',
                    wndmtrx=sample(c(TRUE, FALSE), 1))
   expect_that(tree['ntips'], equals(4))
```
Comment (DomBennett) [inline tests/testthat/test-rw-methods.R:41]: Can you add this? Probably should have been there all along.
```suggestion
  expect_equal(tree['othr_slt_nms'], 'bootstrap')
  expect_true(all(c('0.9', '0.8') %in%
                    as.character(getNdsSlt(tree = tree, slt_nm = 'bootstrap',
                                           ids = tree['nds']))))
  expect_that(tree['ntips'], equals(4))
```

File: R/cnvrt-methods.R [hunk 8]
```diff
@@ -50,7 +50,16 @@ setAs(from="phylo", to="TreeMan", def=function(from, to) {
 #' tree <- as(tree, 'TreeMan')
 setAs(from="TreeMan", to="phylo", def=function(from, to) {
```
Comment (DomBennett) [inline R/cnvrt-methods.R:52]: Can you add something along the lines of "will attempt to preserve internal node id labels" or similar to the @description sections of the conversion docs? Thanks.

**Review by DomBennett — CHANGES_REQUESTED** [review]
Wow, thanks for the contribution!

Please consider the following changes that I've suggested and I'll approve it.

File: R/cnvrt-methods.R (hunk 2 above)
Comment (T-Heide) [reply R/cnvrt-methods.R]: Definitely not ...

File: tests/testthat/test-rw-methods.R (hunk 6 above)
Comment (T-Heide) [reply tests/testthat/test-rw-methods.R]: I have split the test into 'readTree([w/ special node slot]) works' and 'readTree([w/ non-standard node labels])' now.

**Comment by T-Heide** [comment]
That should resolve all the comments.

**Comment by DomBennett** [comment]
Nice, thanks!

**Review by DomBennett — COMMENTED** [review]
Accepted!

**Pull request merged**